runtime/src/bpf_helper.cpp

```diff
@@ -64,20 +66,93 @@ long bpftime_strncmp(const char *s1, uint64_t s1_sz, const char *s2)
 	return strncmp(s1, s2, s1_sz);
 }
 
+static sigjmp_buf jump_buffer;
+
+// used for sigsetjmp to avoid faulting in probe_write_user & probe_read_user
+void segv_handler(int signum)
+{
+	siglongjmp(jump_buffer, 1);
+}
+
 uint64_t bpftime_probe_read(uint64_t dst, uint64_t size, uint64_t ptr, uint64_t,
 			    uint64_t)
 {
-	memcpy((void *)(uintptr_t)dst, (void *)(uintptr_t)ptr,
-	       (size_t)(uint32_t)(size));
-	return 0;
+	if (size <= 0) {
+		return -EINVAL;
+	}
+	struct sigaction sa, old_sa;
+	uint64_t ret = 0;
+
+	sa.sa_handler = segv_handler; // set signal handler
+	sigemptyset(&sa.sa_mask); // clear signal set
+	sa.sa_flags = 0;
+
+	if (sigaction(SIGSEGV, &sa, &old_sa) == -1) {
```

Comment:
Do we have any method to optimize these syscalls? For now, each call to bpftime_probe_read may lead to a series syscalls, which are relatively slow. 

Maybe we can setup signal handlers only once, and in signal handler, walk stack trace to dertermine whether a call to bpftime_probe_read triggers this signal, and determine to let the call fail or crash.
Will this be too hard?

Reply:
I will make a try for this idea later.

Reply:
You can try to use https://github.com/eunomia-bpf/bpftime/blob/master/benchmark%2FREADME.md to see how it performs.